README.md
lua-libmodbus
=============

Lua bindings to [libmodbus](http://www.libmodbus.org/)

Parameters are mostly as per libmodbus documentation, but values are returned
directly, in tables, rather than in pointers and return codes.  Instead of
a return code, you would get a lua style "nil, error_msg" return pair.

Of particular relevance to modbus, is that tables are addressed with lua
style 1 based counting, but converted to zero based internally.  ie:

```
	res = dev:read_registers(0x2000, 2)
	print(#res) -- prints 2
	print(res[1]) - prints register address 0x2000, _not_ 0x2001
```

[API documentation](http://etactica.github.io/lua-libmodbus/) (Generated from last release)

Status
------

* Client bindings for RTU/TCP and almost all operations.
* Some helpers for working with 16/32bit signed/unsigned and floats in multiple registers
  (API is not necessarily nailed down, comments welcome)
* Server side limited to receive and reply exception.
  Needs thoughts on how to handle the mapping objects.
* Compatible with both 3.0.x and 3.1.x but you must run with the version you compiled with.

Compile
-------
You need Lua and libmodbus development packages (headers and libs) to
build lua-libmodbus.

Compile with

    make

You can override the pkg-config package name to set a specific Lua version.
For example:

    make LUAPKG=lua-5.1

Example usage
-------------

See demo.lua and other sample files.

Here is a simple example printing out 10 registers at 0x2000,
then writing 5 registers back to that address

```Lua
mb = require("libmodbus")
local dev = mb.new_tcp_pi("192.168.255.74", 1502)
local base_address = 0x2000
dev:connect()
local regs, err = dev:read_registers(base_address, 10)
if not regs then error("read failed: " .. err) end
for r,v in ipairs(regs) do
        print(string.format("register (offset %d) %d: %d (%#x): %#x (%d)",
                r, r, r + base_address - 1, r + base_address -1, v, v))
end
```

Another example writing registers instead

```Lua
mb = require("libmodbus")
local dev = mb.new_tcp_pi("192.168.255.74", 1502)
dev:connect()

-- will autoround 32.98 to 32.  no magic support for writing floats yet :|
-- numbers are cast to uint16_t internally.
local res, err = dev:write_registers(0x2000, { 0xabcd, 32.98, 0xfffe, 0xabcd, -1 })
if not res then print(err) end
```


Testing
-------
Some spec files for use with busted are provided.  Some of them expect to have
real hardware to talk to, so you may wish to disable them.

```
busted --exclude-tags=real
```


changelog
0.8 2022 November
Add modbus_rtu_{get,set}_rts
Add modbus_rtu_{get,set}_rts_delay

0.7 2021 February
Fix helpers.set_s32 for "lnum" patch environments. #12

0.6.1 2020 July
No functional change: tag acrobatics for luarocks+releases at the same time

0.6 2020 July
add modbus_rtu_{set,get}_serial_mode

0.5 2019 August
Fix compilation under windows/mingw

0.4.2: 2018 November
Fix bad free when closing RTU contexts. #5
Fix compat with lua 5.3 when backwards compat is turned off #6

0.4.1: 2018 June
Drop unintended debug
Fix documentation

0.4: 2018 June
Add number helpers for dealing with 16/32bit signed/unsigned
Add tostring() functionality
Add documentation

0.3.1: 2017 March
Support building with musl libc

0.3: 2017 March
Add support for send_raw_request

0.2: 2016-09-09
Add explicit connection close() method

0.1: Initial release
Basic overage of libmodbus api


spec/modbus_spec.lua
--[[
--You'll need a working modbus device of some description for these tests...
--that's.... out of scope....
--]]
local D = {
	slave = 0xb6,
	host = "192.168.255.74",
	service = 1502,
	base = 0x2000,
	count = 10,
}

require("busted")
mb = require("libmodbus")

describe("basic arg passing", function()

	it("should accept helpful new_ params", function()
		assert.truthy(mb.new_tcp_pi("192.168.255.74", "mbap"))
		assert.truthy(mb.new_tcp_pi("192.168.255.74", 1502))
		assert.has_error(function() mb.new_tcp_pi(true, "xxx") end)
		assert.has_error(function() mb.new_tcp_pi("xxx", {}) end)
	end)

	it("should get/set byte timeouts properly", function()
		x = mb.new_tcp_pi("blah", 123)
		local tv = {x:get_byte_timeout()}
		assert.is_same({0, 500*1000}, tv)
		x:set_byte_timeout(1, 250*1000)
		tv = {x:get_byte_timeout()}
		assert.is_same({1, 250*1000}, tv)
	end)
	it("should get/set response timeouts properly", function()
		x = mb.new_tcp_pi("blah", 123)
		local tv = {x:get_response_timeout()}
		assert.is_same({0, 500*1000}, tv)
		x:set_response_timeout(1, 250*1000)
		tv = {x:get_response_timeout()}
		assert.is_same({1, 250*1000}, tv)
	end)

end)

describe("functional tcp pi tests #real", function()
	local x
	setup(function()
		x = mb.new_tcp_pi(D.host, D.service)
	end)

	it("should connect ok", function()
		assert.is_truthy(x:connect())
		assert.is_truthy(x:set_slave(D.slave))
		local res, err = x:read_registers(D.base, D.count)
		assert.is_truthy(res, err)
	end)

	it("should fail well", function()
		assert.is_truthy(x:connect())
		assert.is_truthy(x:set_slave(D.slave))
		local res, err = x:read_registers(9999, 8)
		assert.falsy(res, "should have failed with illegal address")
	end)
		
	
end)

spec/number_helpers_spec.lua
require("busted")
--local rmep = require("remake.plugin_support")
local mb = require("libmodbus")

function noit()
end

function nodescribe()
end

describe("get signed vs unsigned", function()
    it("should decode unsigned", function()
        regs = { 0x0020, 0x3456 }
        val = mb.get_s32(unpack(regs))
        valu = mb.get_u32(unpack(regs))
        assert.are.equal(val, valu)
        assert.are.equal(val, 2110550)
    end)
    it("shouldn't hurt things", function()
	assert.are.equal(5, mb.get_s16(5))
	assert.are.equal(-5, mb.get_s16(-5))
	assert.are.equal(-5, mb.get_s16(0xfffb))
	assert.are.equal(-5, mb.get_s16(65531))
	-- This isn't reasonable.  In lua's view, 0xface is a moderately large number.
	-- it's only in modbus that we might consider this to be a negative 16bit.
	-- assert.are.equal(0xface, mb.get_s16(0xface))
	assert.are.equal(0x1ace, mb.get_s16(0x1ace))
    end)
    it("should decode signed negative and large positive", function()
        regs = { 0xffff, 0xfc44 }
        val = mb.get_s32(unpack(regs))
        valu = mb.get_u32(unpack(regs))
        assert.are.equal(-956, val)
        -- not all lua's interpret hex literals the same way, and this form is only for visual sanity
        --assert.are.equal(0xfffffc44, valu)
        assert.are.equal(4294966340, valu)
    end)
    it("should do near edges", function()
        regs = { 0xffff, 0xfffe }
        assert.are.equal(-2, mb.get_s32(unpack(regs)))
        valu = mb.get_u32(unpack(regs))
        -- not all lua's interpret hex literals the same way, and this form is only for visual sanity
        --assert.are.equal(0xfffffffe, valu)
        assert.are.equal(4294967294, valu)
    end)
    it("should do edges_high", function()
        regs = { 0xffff, 0xffff }
        assert.are.equal(-1, mb.get_s32(unpack(regs)))
        assert.are.equal(4294967295, mb.get_u32(unpack(regs)))
    end)
    it("should do edges_low", function()
        regs = { 0x0, 0x0 }
        assert.are.equal(0, mb.get_s32(unpack(regs)))
        assert.are.equal(0, mb.get_u32(unpack(regs)))
    end)
end)


describe("set signed vs unsigned", function()
	it("should set signed nicely", function()
		local h,l = mb.set_s32(2110550)
		assert.are.equal(0x0020, h)
		assert.are.equal(0x3456, l)
	end)
	it("should set signed negative", function()
		local h,l = mb.set_s32(-2110550)
		-- different lua implementations treat literal hex differently.
		assert.are.equal(mb.get_s16(0xffdf), mb.get_s16(h))
		assert.are.equal(mb.get_s16(0xcbaa), mb.get_s16(l))
	end)
end)

describe("floating point should do sane things", function()
	it("simple positive truncation", function()
		local v = 5698.123
		local h,l = mb.set_s32(v)
		assert.are.equal(0, h)
		assert.are.equal(0x1642, l)
		assert.are.equal(5698, l)
	end)
	it("simple negative truncation", function()
		local v = -5698.123
		local h,l = mb.set_s32(v)
		assert.are.equal(mb.get_s16(0xffff), mb.get_s16(h))
		assert.are.equal(mb.get_s16(0xe9be), mb.get_s16(l))
		assert.are.equal(-5698, mb.get_s16(l))
	end)
	it("bracketed scaling", function()
		local v = 5698.123
		local scale = 1000
		local h,l = mb.set_s32(v*scale)
		assert.are.equal(0x56, h)
		assert.are.equal(mb.get_s16(0xf24b), mb.get_s16(l))
	end)
	it("bracketed negative scaling", function()
		local v = -5698.123
		local scale = 1000
		local h,l = mb.set_s32(v*scale)
		assert.are.equal(mb.get_s16(0xffa9), mb.get_s16(h))
		assert.are.equal(0x0db5, l)
	end)
end)

describe("regswap", function()
    it("should work normally", function()
        regs = {0x1122, 0x3344}
        assert.are_equal(860098850, mb.get_s32le(unpack(regs)))
        assert.are_equal(287454020, mb.get_s32(unpack(regs)))
        assert.are_equal(860098850, mb.get_u32le(unpack(regs)))
        assert.are_equal(287454020, mb.get_u32(unpack(regs)))
    end)
    it("should handle negatives too", function()
        regs = { 0xfffa, 0xfffe }
        -- not all lua's interpret hex literals the same way, and this form is only for visual sanity
        --assert.are.equal(0xfffefffa, mb.get_u32le(unpack(regs)))
        assert.are.equal(4294901754, mb.get_u32le(unpack(regs)))
        assert.are.equal(-65542, mb.get_s32le(unpack(regs)))
    end)
end)

describe("64bit ops", function()
    it("should handle basic stuff", function()
        regs = {0xffff, 0xffff, 0xffff, 0xfffe}
        local valu = mb.get_u64(unpack(regs))
        assert.are.equal(18446744073709551614, mb.get_u64(unpack(regs)))
        assert.are.equal(-2, mb.get_s64(unpack(regs)))
    end)
    it("should handle small stuff too", function()
        regs = {0x0001, 0x0001, 0xffff, 0xfffe}
        assert.are.equal(281483566645246, mb.get_s64(unpack(regs)))
        assert.are.equal(281483566645246, mb.get_u64(unpack(regs)))
    end)
    it("should handle zero", function()
        regs = {0x0000, 0x0000, 0x0000, 0x0000}
        assert.are.equal(0, mb.get_s64(unpack(regs)))
        assert.are.equal(0, mb.get_u64(unpack(regs)))
    end)
    it("should handle one", function()
        regs = {0x0000, 0x0000, 0x0000, 0x0001}
        assert.are.equal(1, mb.get_s64(unpack(regs)))
        assert.are.equal(1, mb.get_u64(unpack(regs)))
    end)
    it ("should handle minus one", function()
        regs = {0xffff, 0xffff, 0xffff, 0xffff}
        assert.are.equal(-1, mb.get_s64(unpack(regs)))
    end)
    it ("should handle giant number", function()
        regs = {0xffff, 0xffff, 0xffff, 0xffff}
        local result = mb.get_u64(unpack(regs))
        assert.is_true(result >= 1.84467440737E19 and result <= 1.844674407372E19)
    end)
    it ("should handle giant negative number", function()
        regs = {0x8000, 0x0000, 0x0000, 0x0000}
        local result = mb.get_s64(unpack(regs))
        assert.is_true(result >= -9.2233720368549E18 and result <= -9.2233720368547E18)
    end)
end)


demo-rtu.lua
--[[
Very simple demo of using client/master APIs
Karl Palsson, April 2016
--]]

mb = require("libmodbus")
print("using libmodbus runtime version: ", mb.version())
print("using lua-libmodbus compiled against libmodbus: ", mb.VERSION_STRING)

-- everything but device will default to modbus recommendations.
local dev = mb.new_rtu("/dev/ttyUSB0", 19200, "even", 8, 1)
print("byte timeout:", dev:get_byte_timeout())
print("response timeout", dev:get_response_timeout())
dev:set_debug()
--dev:rtu_set_serial_mode(mb.RTU_RS232)
--dev:set_error_recovery(mb.ERROR_RECOVERY_LINK, mb.ERROR_RECOVERY_PROTOCOL)
--dev:set_error_recovery(mb.ERROR_RECOVERY_NONE)
ok, err = dev:connect()
if not ok then error("Couldn't connect: " .. err) end

function simple_read(dev)
	dev:set_slave(0x88)
	local base_address = 0x2000
	local regs, err
	regs, err = dev:read_registers(base_address, 10)
	if not regs then error("read failed: " .. err) end

	for r,v in ipairs(regs) do
		print(string.format("register (offset %d) %d: %d (%#x): %#x (%d)",
			r, r, r + base_address - 1, r + base_address -1, v, v))
	end

	--report = dev:report_slave_id()
	--print("report slaveid: ", report:byte(1), report:byte(2), report:sub(3))

end
	
simple_read(dev)



demo.lua
--[[
Very simple demo of using client/master APIs
Karl Palsson, April 2016
--]]

mb = require("libmodbus")
print("using libmodbus runtime version: ", mb.version())
print("using lua-libmodbus compiled against libmodbus: ", mb.VERSION_STRING)

local dev = mb.new_tcp_pi("localhost", 1502)
print(dev:get_byte_timeout())
print(dev:get_response_timeout())
print(dev)
dev:set_debug()
--dev:set_error_recovery(mb.ERROR_RECOVERY_LINK, mb.ERROR_RECOVERY_PROTOCOL)
--dev:set_error_recovery(mb.ERROR_RECOVERY_NONE)
ok, err = dev:connect()
if not ok then error("Couldn't connect: " .. err) end

function lmb_unit_test(dev)
	dev:set_slave(17)
	local base_address = 0x16b
	local regs, err
	regs, err = dev:read_registers(base_address, 3)
	if not regs then error("read failed: " .. err) end

	print("read 3 from 0x16b: ")
	for r,v in ipairs(regs) do
		print(string.format("register (offset %d) %d: %d (%#x): %#x (%d)",
			r, r, r + base_address - 1, r + base_address -1, v, v))
	end

	print("write single 0x16c -> 0xabcd ", dev:write_register(0x16c, 0xabcd))
	regs, err = dev:read_registers(0x16c, 1)
	if not regs then error("read1 failed: ", err) end
	print("reading back:  single 0x16c: ", string.format("%x", regs[1]))

	regs, err = dev:read_input_registers(0x108, 1)
	if not regs then print("read input regs failed", err) end
	print("should be 0xa: ", string.format("%x", regs[1]))

	bits, err = dev:read_bits(0x130, 0x25)
	if not bits then print("read bits failed", err) end
	for r,v in ipairs(bits) do
		print(string.format("bit (offset %d) %d: %d (%#x): %#x (%d)",
			r, r, r + 0x130 - 1, r + 0x130 -1, v, v))
	end

	bits, err = dev:write_bits(0x130, { 0xffff,0, true,0,false,1,0,99 })
	if not bits then print("write bits failed", err) end
	bits, err = dev:read_bits(0x130, 8)
	print("reading back 8 bits, should be 0xa5")
	for r,v in ipairs(bits) do
		print(string.format("bit (offset %d) %d: %d (%#x): %#x (%d)",
			r, r, r + 0x130 - 1, r + 0x130 -1, v, v))
	end

	err = dev:write_bit(0x131, 1);
	if err then print("write bit returned:", err) end
	bits, err = dev:read_bits(0x131, 1)
	if not bits then print("re-reading bits failed", err) end
	print("rereading bits should be 1 now: ", bits[1])

	bits, err = dev:read_input_bits(0x1c4, 8)
	if not bits then print("read bits failed", err) end
	for r,v in ipairs(bits) do
		print(string.format("bit (offset %d) %d: %d (%#x): %#x (%d)",
			r, r, r + 0x1c4 - 1, r + 0x1c4 -1, v, v))
	end

	report = dev:report_slave_id()
	print("report slaveid: ", report:byte(1), report:byte(2), report:sub(3))

end
	

function write_test(dev)
-- will autoround 32.98 to 32.  no magic support for writing floats yet :|
-- numbers are cast to uint16_t internally.
	local res, err = dev:write_registers(0x2000, { 0xabcd, 32.98, 0xfffe, 0xabcd, -1 })
	if not res then print(err) end
	regs, err = dev:read_registers(base_address, 5)
	print("so, no magic casting 16bit tosigned!", regs[5], -1)

	print(dev:write_register(0x2006, 2))
end

lmb_unit_test(dev)
